runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -214,21 +213,22 @@ State waitUntilFinish(
     MonitoringUtil monitor = new MonitoringUtil(projectId, dataflowOptions.getDataflowClient());
 
     long lastTimestamp = 0;
-    BackOff backoff =
-        duration.getMillis() > 0
-            ? new AttemptAndTimeBoundedExponentialBackOff(
-                MESSAGES_POLLING_ATTEMPTS,
-                MESSAGES_POLLING_INTERVAL,
-                duration.getMillis(),
-                AttemptAndTimeBoundedExponentialBackOff.ResetPolicy.ATTEMPTS,
-                nanoClock)
-            : new AttemptBoundedExponentialBackOff(
-                MESSAGES_POLLING_ATTEMPTS, MESSAGES_POLLING_INTERVAL);
+    FlexibleBackoff backoff = FlexibleBackoff.of()
+        .withMaxRetries(MESSAGES_POLLING_RETRIES)
+        .withInitialBackoff(MESSAGES_POLLING_INTERVAL);
+    if (duration.isLongerThan(Duration.ZERO)) {
+      backoff = backoff.withMaxCumulativeBackoff(duration);
+    }
+
+    // In this function, track the cumulative time since we've started backing off and use this
+    // to enforce a wall-clock limit.
+    long startNanos = nanoClock.nanoTime();
```

Comment:
Maybe clarify the comment to state that the backoff utility doesn't allow for limiting time from first-attempt to success? Should it?

Or is this time from first-attempt at first request to success of all requests (where each request is separately retried)?